File: src/memory-maintenance.ts
/**
 * Memory maintenance for decay, promotion, and reflection generation.
 *
 * Operations:
 * - Memory decay: importance *= 0.98 daily
 * - Memory promotion: episodic -> semantic when access_count > 10
 * - Reflection generation: every 50 episodic memories
 */

import { SurrealDatabase } from './surrealdb-client.js';
import { ImportanceLearning } from './importance-learning.js';
import { logInfo, logError } from './maintenance-logger.js';

export interface MemoryMaintenanceConfig {
  decayFactor?: number;
  promotionThreshold?: number;
  reflectionInterval?: number;
  olderThanDays?: number;
}

export class MemoryMaintenance {
  private static readonly DECAY_FACTOR = 0.98;
  private static readonly PROMOTION_THRESHOLD = 10;
  private static readonly REFLECTION_INTERVAL = 50;

  private config: Required<MemoryMaintenanceConfig>;
  private importanceLearning: ImportanceLearning;

  constructor(
    private db: SurrealDatabase,
    config: MemoryMaintenanceConfig = {}
  ) {
    this.config = {
      decayFactor: config.decayFactor ?? MemoryMaintenance.DECAY_FACTOR,
      promotionThreshold: config.promotionThreshold ?? MemoryMaintenance.PROMOTION_THRESHOLD,
      reflectionInterval: config.reflectionInterval ?? MemoryMaintenance.REFLECTION_INTERVAL,
      olderThanDays: config.olderThanDays ?? 1,
    };
    this.importanceLearning = new ImportanceLearning();
  }

  /**
   * Run decay on memories older than specified days.
   * Returns the number of memories decayed.
   */
  async runDecay(): Promise<{ episodicDecayed: number; semanticDecayed: number }> {
    const cutoff = new Date();
    cutoff.setDate(cutoff.getDate() - this.config.olderThanDays);
    const cutoffStr = cutoff.toISOString();

    let episodicDecayed = 0;
    let semanticDecayed = 0;

    // Scroll through all memories and apply decay
    for (const type of ['episodic', 'semantic'] as const) {
      let offset = 0;
      while (true) {
        const memories = await this.db.scroll({ type }, 100, offset);
        if (memories.length === 0) break;

        for (const mem of memories) {
          const createdAt = new Date(mem.payload.created_at);
          if (createdAt < cutoff) {
            // Apply decay
            const oldImportance = mem.payload.importance || 0.5;
            const newImportance = Math.max(0.01, oldImportance * this.config.decayFactor);
            await this.db.updatePayload(mem.id, {
              ...mem.payload,
              importance: newImportance,
            });

            if (type === 'episodic') episodicDecayed++;
            else semanticDecayed++;
          }
        }

        offset += memories.length;
        if (memories.length < 100) break;
      }
    }

    return { episodicDecayed, semanticDecayed };
  }

  /**
   * Check for memories to promote (episodic -> semantic).
   * Returns the number of memories promoted.
   */
  async runPromotion(): Promise<number> {
    let promoted = 0;
    let offset = 0;

    while (true) {
      const memories = await this.db.scroll({ type: 'episodic' }, 100, offset);
      if (memories.length === 0) break;

      for (const mem of memories) {
        const accessCount = mem.payload.access_count || 0;
        if (accessCount >= this.config.promotionThreshold) {
          // Promote to semantic
          const content = mem.payload.content || mem.payload.summary || '';
          await this.db.upsert(mem.id, [], {
            type: 'semantic',
            content: content,
            importance: mem.payload.importance,
            access_count: accessCount,
            created_at: mem.payload.created_at,
          });

          promoted++;
        }
      }

      offset += memories.length;
      if (memories.length < 100) break;
    }

    return promoted;
  }

  /**
   * Generate reflection if enough episodic memories exist.
   * Returns true if a reflection was generated.
   */
  async maybeGenerateReflection(
    generateFn: (summaries: string[]) => Promise<string>
  ): Promise<boolean> {
    // Count episodic memories by scrolling through them
    let episodicCount = 0;
    let offset = 0;

    while (true) {
      const memories = await this.db.scroll({ type: 'episodic' }, 100, offset);
      if (memories.length === 0) break;
      episodicCount += memories.length;
      offset += memories.length;
    }

    // Check if count is a multiple of reflection interval
    if (episodicCount < this.config.reflectionInterval) {
      return false;
    }

    if (episodicCount % this.config.reflectionInterval !== 0) {
      return false;
    }

    logInfo(`[MemoryMaintenance] Generating reflection: ${episodicCount} episodic memories (interval: ${this.config.reflectionInterval})`);

    try {
      // Get recent episodic memories for summarization
      const recentMemories: string[] = [];
      let memOffset = 0;

      while (recentMemories.length < this.config.reflectionInterval) {
        const memories = await this.db.scroll({ type: 'episodic' }, 50, memOffset);
        if (memories.length === 0) break;

        for (const mem of memories) {
          const content = mem.payload.content || mem.payload.summary || '';
          if (content.trim()) {
            recentMemories.push(content);
          }
        }
        memOffset += memories.length;
      }

      if (recentMemories.length < 10) {
        logInfo('[MemoryMaintenance] Not enough content for reflection');
        return false;
      }

      // Generate reflection using LLM
      const reflection = await generateFn(recentMemories.slice(0, 50));

      if (reflection && reflection.trim().length > 0 && reflection !== 'No significant content.') {
        logInfo(`[MemoryMaintenance] Generated reflection: "${reflection.substring(0, 50)}..."`);
        return true;
      }

      return false;
    } catch (error: any) {
      logError(`[MemoryMaintenance] Reflection generation failed: ${error.message}`);
      return false;
    }
  }

  /**
   * Run all maintenance tasks.
   */
  async runMaintenance(
    generateReflectionFn?: (summaries: string[]) => Promise<string>
  ): Promise<{
    episodicDecayed: number;
    semanticDecayed: number;
    promoted: number;
    reflectionGenerated: boolean;
  }> {
    const [decayResult, promoted] = await Promise.all([
      this.runDecay(),
      this.runPromotion(),
    ]);

    let reflectionGenerated = false;
    if (generateReflectionFn) {
      reflectionGenerated = await this.maybeGenerateReflection(generateReflectionFn);
    }

    return {
      ...decayResult,
      promoted,
      reflectionGenerated,
    };
  }
}

File: memory_maintenance.py
"""Memory maintenance for decay, promotion, and reflection generation."""

import logging
from datetime import datetime, timedelta
from typing import Callable, Optional, List, Dict

from database import Database
from episodic_memory import EpisodicMemory, EpisodicMemoryStore
from semantic_memory import SemanticMemory, SemanticMemoryStore
from reflection_memory import ReflectionMemory, ReflectionMemoryStore
from importance_learning import ImportanceLearning

logger = logging.getLogger(__name__)


class MemoryMaintenance:
    """Handles memory maintenance operations.

    Operations:
    - Memory decay: importance *= 0.98 daily
    - Memory promotion: episodic -> semantic when access_count > 10
    - Reflection generation: every 50 episodic memories
    """

    # Decay factor (daily)
    DECAY_FACTOR = 0.98

    # Promotion threshold
    PROMOTION_THRESHOLD = 10

    # Reflection interval (number of episodic memories)
    REFLECTION_INTERVAL = 50

    def __init__(
        self,
        db: Database,
        llm_summarize_fn: Optional[Callable[[List[str]], str]] = None
    ) -> None:
        """Initialize memory maintenance.

        Args:
            db: Database instance.
            llm_summarize_fn: Optional function to generate summaries using LLM.
        """
        self.db = db
        self.episodic_store = EpisodicMemoryStore(db)
        self.semantic_store = SemanticMemoryStore(db)
        self.reflection_store = ReflectionMemoryStore(db)
        self.importance_learner = ImportanceLearning()
        self.llm_summarize_fn = llm_summarize_fn

    def run_decay(self, older_than_days: int = 1) -> int:
        """Apply decay to memories older than specified days.

        Args:
            older_than_days: Only decay memories older than this.

        Returns:
            Number of memories decayed.
        """
        cutoff = datetime.now() - timedelta(days=older_than_days)

        # Get old episodic memories
        sql = """
            SELECT id, importance, created_at
            FROM episodic_memory
            WHERE created_at < %s
        """
        results = self.db.query(sql, (cutoff,))

        count = 0
        for row in results:
            new_importance = self.importance_learner.apply_decay(
                row["importance"],
                self.DECAY_FACTOR
            )
            update_sql = """
                UPDATE episodic_memory
                SET importance = %s
                WHERE id = %s
            """
            self.db.execute(update_sql, (new_importance, row["id"]))
            count += 1

        # Get old semantic memories
        sql = """
            SELECT id, importance, created_at
            FROM semantic_memory
            WHERE created_at < %s
        """
        results = self.db.query(sql, (cutoff,))

        for row in results:
            new_importance = self.importance_learner.apply_decay(
                row["importance"],
                self.DECAY_FACTOR
            )
            update_sql = """
                UPDATE semantic_memory
                SET importance = %s
                WHERE id = %s
            """
            self.db.execute(update_sql, (new_importance, row["id"]))
            count += 1

        logger.info(f"Applied decay to {count} memories")
        return count

    def run_promotion(self) -> int:
        """Promote episodic memories to semantic based on access count.

        Returns:
            Number of memories promoted.
        """
        # Find episodic memories with high access count
        sql = """
            SELECT id, content, importance, access_count
            FROM episodic_memory
            WHERE access_count > %s
        """
        results = self.db.query(sql, (self.PROMOTION_THRESHOLD,))

        count = 0
        for row in results:
            # Check if already promoted (content exists in semantic)
            check_sql = """
                SELECT id FROM semantic_memory
                WHERE content = %s
            """
            existing = self.db.query(check_sql, (row["content"],))

            if not existing:
                # Promote to semantic memory
                self.semantic_store.promote_from_episodic(
                    content=row["content"],
                    importance=row["importance"]
                )
                count += 1
                logger.info(f"Promoted episodic memory {row['id']} to semantic")

        logger.info(f"Promoted {count} memories to semantic")
        return count

    def check_and_generate_reflection(self) -> Optional[int]:
        """Check if reflection should be generated and generate if needed.

        Returns:
            Reflection memory ID if generated, None otherwise.
        """
        total_count = self.episodic_store.get_count()

        # Check if we've hit reflection interval
        if total_count > 0 and total_count % self.REFLECTION_INTERVAL == 0:
            return self.generate_reflection()

        return None

    def generate_reflection(self) -> Optional[int]:
        """Generate a reflection memory from recent episodic memories.

        Returns:
            Reflection memory ID if successful, None otherwise.
        """
        # Get recent memories
        memories = self.episodic_store.get_all_for_reflection(
            limit=self.REFLECTION_INTERVAL
        )

        if not memories:
            return None

        # Extract content for summarization
        contents = [m.content for m in memories]

        # Generate summary using LLM if available
        if self.llm_summarize_fn:
            try:
                summary = self.llm_summarize_fn(contents)
            except Exception as e:
                logger.error(f"Failed to generate reflection summary: {e}")
                summary = self._fallback_summary(contents)
        else:
            summary = self._fallback_summary(contents)

        # Store reflection
        reflection = ReflectionMemory(summary=summary, importance=0.9)
        reflection_id = self.reflection_store.store(reflection)

        logger.info(f"Generated reflection memory {reflection_id}")
        return reflection_id

    def _fallback_summary(self, contents: List[str]) -> str:
        """Generate a simple fallback summary without LLM.

        Args:
            contents: List of memory contents.

        Returns:
            Simple summary string.
        """
        # Just list the topics covered
        unique_topics = list(set(contents))[:10]
        return f"Key experiences: {'; '.join(unique_topics)}"

    def run_maintenance(self) -> Dict:
        """Run all maintenance operations.

        Returns:
            Dict with counts of operations performed.
        """
        results = {
            "decayed": self.run_decay(),
            "promoted": self.run_promotion(),
            "reflections_generated": 0
        }

        reflection_id = self.check_and_generate_reflection()
        if reflection_id:
            results["reflections_generated"] = 1

        return results
